extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,283 @@
+    this.isReusable_ = false;
+
+    // Register default error modes.
+    for (const failureType in FailureTypes) {
+      this.recoveryModes_[failureType] = RecoveryModeTypes.COLLAPSE;
+    }
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.FailureType} failureType
+   * @param {!./amp-ad-type-defs.RecoveryModeType} recoveryType
+   * @param {number=} retryTimer
+   * @param {string=} fallback Validator/Renderer fallback.
+   * @final
+   */
+  onFailure(failureType, recoveryType, retryTimer, fallback) {
+    if (this.recoveryModes_[failureType]) {
+      dev().warn(TAG,
+          `Recovery mode for failure type ${failureType} already registered!`);
+    }
+    this.checkAndSetRecoveryMode_(
+        failureType, {type: recoveryType, retryTimer, fallback});
+  }
+
+  /**
+   * @param {!./amp-ad-render.Validator} validator
+   * @param {string=} type
+   * @final
+   */
+  registerValidator(validator, type = 'default') {
+    if (this.validators_[type]) {
+      dev().warn(TAG, `${type} validator already registered.`);
+    }
+    this.validators_[type] = validator;
+  }
+
+  /**
+   * @param {!./amp-ad-render.Renderer} renderer
+   * @param {string} type
+   * @final
+   */
+  registerRenderer(renderer, type) {
+    if (this.renderers_[type]) {
+      dev().warn(TAG, `Rendering mode already registered for type '${type}'`);
+    }
+    this.renderers_[type] = renderer;
+  }
+
+  /**
+   * @return {string} The finalized ad request URL.
+   * @protected
+   * @abstract
+   */
+  getRequestUrl() {
+    // Subclass must override.
+  }
+
+  /** @param {boolean} isReusable */
+  setIsReusable(isReusable) {
+    this.isReusable_ = isReusable;
+  }
+
+
+  /**
+   * Processes the ad response as soon as the XHR request returns. This can be
+   * overridden and used as a hook to perform any desired logic before passing
+   * the response to the validator.
+   * @param {?../../../src/service/xhr-impl.FetchResponse} response
+   * @private
+   */
+  handleAdResponse_(response) {
+    if (!response.arrayBuffer) {
+      this.handleFailure_(FailureTypes.NO_RESPONSE);
+      return;
+    }
+    response.arrayBuffer().then(unvalidatedBytes => {
+      const validatorType = response.headers.get('validator-type') || 'default';
+      this.context_
+          .setUnvalidatedBytes(unvalidatedBytes)
+          .setHeaders(response.headers);
+      this.invokeValidator_(validatorType);
+    });
+  }
+
+  /**
+   * @param {string} validatorType
+   * @private
+   */
+  invokeValidator_(validatorType) {
+    const unvalidatedBytes = this.context_.getUnvalidatedBytes();
+    dev().assert(this.validators_[validatorType],
+        'Validator never registered!');
+    dev().assert(unvalidatedBytes,
+        'Validator invoked before ad response received!');
+    this.validators_[validatorType].validate(this.context_)
+        .then(context => {
+          this.handleValidatorResponse_(context);
+          if (!this.isReusable_) {
+            this.validators_ = map({});
```

Comment:
any specific reason why we clean up validators?

Reply:
@jridgewell requested it.

Reply:
@jridgewell what was the concern? There will be likely just 1-2 validators on the page. 
Any other example in AMP that we are so eagerly recycling a map?